src/Authenticator/CookieAuthenticator.php

```diff
@@ -0,0 +1,208 @@
+<?php
+/**
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Authentication\Authenticator;
+
+use Authentication\Identifier\IdentifierCollection;
+use Authentication\PasswordHasher\PasswordHasherTrait;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use RuntimeException;
+
+/**
+ * Cookie Authenticator
+ *
+ * Authenticates an identity based on a cookies data.
+ */
+class CookieAuthenticator extends AbstractAuthenticator implements PersistenceInterface
+{
+
+    use PasswordHasherTrait;
+
+    /**
+     * {@inheritDoc}
+     */
+    protected $_defaultConfig = [
+        'rememberMeField' => 'remember_me',
+        'fields' => [
+            'username' => 'username',
+            'password' => 'password'
+        ],
+        'cookie' => [
+            'name' => 'CookieAuth',
+            'expire' => null,
+            'path' => '/',
+            'domain' => '',
+            'secure' => false,
+            'httpOnly' => false
+        ],
+        'passwordHasher' => 'Authentication.Default'
+    ];
+
+    /**
+     * {@inheritDoc}
+     */
+    public function __construct(IdentifierCollection $identifiers, array $config = [])
+    {
+        $this->_checkCakeVersion();
+
+        parent::__construct($identifiers, $config);
+    }
+
+    /**
+     * Checks the CakePHP Version by looking for the cookie implementation
+     *
+     * @return void
+     */
+    protected function _checkCakeVersion()
+    {
+        if (!class_exists('Cake\Http\Cookie\Cookie')) {
+            throw new RuntimeException('Install CakePHP version >=3.5.0 to use the `CookieAuthenticator`.');
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function authenticate(ServerRequestInterface $request, ResponseInterface $response)
+    {
+        $cookies = $request->getCookieParams();
+        $cookieName = $this->getConfig('cookie.name');
+        if (!isset($cookies[$cookieName])) {
+            return new Result(null, Result::FAILURE_CREDENTIALS_NOT_FOUND, [
+                'Login credentials not found'
+            ]);
+        }
+
+        $token = explode(':', $cookies[$cookieName]);
+
+        if (count($token) !== 2) {
+            return new Result(null, Result::FAILURE_CREDENTIAL_INVALID, [
+                'Cookie token is invalid.'
+            ]);
+        }
+
+        list($username, $tokenHash) = $token;
+
+        $credentials = [
+            'username' => json_decode($username)
```

Comment:
If we json_decode after the split I don't think we'll get around `:` in usernames. I think we'd need to json_encode/decode the entire token to be safe. Something like

```php
$token = json_encode([$username, $hash]);
```

And when checking the token something like:

```php
$value = json_decode($cookies[$cookeName], true);
if ($value === null || count($value) !== 2) {
  // return error
}
```